nixos/modules/services/networking/syncthing.nix

```diff
@@ -54,10 +49,12 @@ let
     old_cfg=$(curl ${cfg.guiAddress}/rest/config)
 
     # generate the new config by merging with the NixOS config options
-    new_cfg=$(printf '%s\n' "$old_cfg" | ${pkgs.jq}/bin/jq -c '. * {
+    new_cfg=$(printf '%s\n' "$old_cfg" | ${pkgs.jq}/bin/jq -c '. * ${builtins.toJSON cfg.settings} * {
         "devices": (${builtins.toJSON devices}${optionalString (cfg.devices == {} || ! cfg.overrideDevices) " + .devices"}),
         "folders": (${builtins.toJSON folders}${optionalString (cfg.folders == {} || ! cfg.overrideFolders) " + .folders"})
-    } * ${builtins.toJSON cfg.extraOptions}')
+    }')
+
+    echo "$new_cfg" | ${pkgs.jq}/bin/jq .
 
```

Comment:
I think the JSON should be shell escaped properly because it might contain `'`. Like so:

```diff
diff --git a/syncthing.nix b/syncthing.nix
index 878969c..1fdda50 100644
--- a/syncthing.nix
+++ b/syncthing.nix
@@ -1,9 +1,7 @@
-    old_cfg=$(curl ${cfg.guiAddress}/rest/config)
+    old_cfg=$(curl -fs ${cfg.guiAddress}/rest/config)
 
     # generate the new config by merging with the NixOS config options
-    new_cfg=$(printf '%s\n' "$old_cfg" | ${pkgs.jq}/bin/jq -c '. * ${builtins.toJSON cfg.settings} * {
+    new_cfg=$(printf '%s\n' "$old_cfg" | ${pkgs.jq}/bin/jq ${escapeShellArg ". * ${builtins.toJSON cfg.settings} * {
         "devices": (${builtins.toJSON devices}${optionalString (cfg.devices == {} || ! cfg.overrideDevices) " + .devices"}),
         "folders": (${builtins.toJSON folders}${optionalString (cfg.folders == {} || ! cfg.overrideFolders) " + .folders"})
-    }')
-
-    echo "$new_cfg" | ${pkgs.jq}/bin/jq .
+    }"})
```

I also added some curl flags and removed formatting of `$new_cfg` in a second step.

Reply:
Should we shellescape folders and devices also?

Reply:
i have no plan about what your talking, so i cant give input to these :D